src/Polysemy/MTL.hs

```diff
@@ -0,0 +1,189 @@
+{-# LANGUAGE AllowAmbiguousTypes         #-}
+{-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}
+
+module Polysemy.MTL
+  (
+  -- * Types
+    CanonicalEffect
+  , ConstrainedAction
+  , ReifiableConstraint1
+  , SemDictionary
+
+  -- * constraint-polymorphic absorber
+  , absorb
+  
+  -- * constraint-monomorphic absorbers
+  , absorbReader
+  , absorbState
+  , absorbWriter
+  , absorbError
+  )
+where
+
+
+import qualified Control.Monad.Reader.Class as S
+import qualified Control.Monad.State.Class as S
+import qualified Control.Monad.Writer.Class as S
+import qualified Control.Monad.Error.Class as S
+import qualified Data.Constraint as C
+import           Data.Constraint ((:-),(\\))
+import qualified Data.Constraint.Unsafe as C
+import           Data.Proxy (Proxy (..))
+import qualified Data.Reflection as R
+import           Data.Kind (Type, Constraint)
+
+import           Polysemy
+import           Polysemy.Reader
+import           Polysemy.Writer
+import           Polysemy.State
+import           Polysemy.Error
+
+{- | Open type-family mapping a single constraint of the form @(Type -> Type) -> Constraint@, e.g., @MonadState s@,
+to a polysemy effect which can be used to re-interpret that constraint, e.g., 'State s'.
+-}
+type family CanonicalEffect (p :: (Type -> Type) -> Constraint) :: (Type -> Type) -> Type -> Type
+
+type instance  CanonicalEffect (S.MonadReader env) = Reader env
+type instance  CanonicalEffect (S.MonadWriter w)   = Writer w
+type instance  CanonicalEffect (S.MonadState s)    = State s
+type instance  CanonicalEffect (S.MonadError e)    = Error e
+
+
+{- | A newtype wrapper for a monadic action, parameterized by
+a constraint, @p@ on a @(Type -> Type)@ (e.g., a monad); @m@, a specific
+@(Type -> Type)@; and a polysemy effect type-list @r@. With "Data.Reflection"
+we can create instances of @p (ConstrainedAction p m r)@ using functions from
+@Sem r@.
+-}
+newtype ConstrainedAction (p :: (Type -> Type) -> Constraint) (m :: Type -> Type) (s :: Type) (x :: Type)
+  = ConstrainedAction
+    { action :: m x
+    } deriving (Functor, Applicative, Monad)
+
+{- | For a constraint to be "absorbable" by @Sem r@, there needs to be an instance of this class,
+containing the dictionary signatures as a record of functions and the
+reflected entailment of @p (ConstrainedAction p m r)@ from the reified dictionary.
+-}
+class ReifiableConstraint1 p where
+  data Dict (p :: (Type -> Type) -> Constraint) (m :: Type -> Type)
+  reifiedInstance :: Monad m => R.Reifies s (Dict p m) :- p (ConstrainedAction p m s)
+
+{- | This class contains an instance of the dictionary for some set of effects
+parameterized by a polysemt effect list @r@.
+Typically, you would write this instance for any @r@
+satisfying the constraint that the "canonical" effect is a member.  But you
+could also use it to discharge constraints which require multiple polysemy effects.
+-}
+class ReifiableConstraint1 p => SemDictionary p r where
```

Comment:
I don't love this name; it feels like the name of the implementation, rather than being insightful. Maybe `IsCanonicalEffect` or something similar?